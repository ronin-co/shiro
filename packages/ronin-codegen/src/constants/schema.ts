import { SyntaxKind, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';

import type { TypeNode } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { DML_QUERY_TYPES } from '../../../ronin-compiler/dist';

/**
 * A list of all model field types & their TypeScript type mapping.
 */
export const MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD = {
  blob: factory.createTypeReferenceNode(identifiers.compiler.storedObject),
  boolean: factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
  date: factory.createTypeReferenceNode(identifiers.primitive.date),
  json: factory.createUnionTypeNode([
    factory.createTypeReferenceNode(identifiers.utils.jsonObject),
    factory.createTypeReferenceNode(identifiers.utils.jsonArray),
  ]),
  link: factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
  number: factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
  string: factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
} satisfies Record<ModelField['type'], TypeNode>;

/**
 * A simple object mapping all DML query types to their human readable string.
 */
export const READABLE_DML_QUERY_TYPES = {
  add: 'Add',
  count: 'Count',
  get: 'Get',
  remove: 'Remove',
  set: 'Set',
} satisfies Record<(typeof DML_QUERY_TYPES)[number], string>;
