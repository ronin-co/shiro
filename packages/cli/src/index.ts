import { parseArgs } from 'node:util';

import path from 'node:path';
import apply from '@/src/commands/apply';
import diff from '@/src/commands/diff';
import initializeProject from '@/src/commands/init';
import logIn from '@/src/commands/login';
import pull from '@/src/commands/pull';
import generateTypes, { TYPES_FLAGS } from '@/src/commands/types';
import { printHelp, printVersion } from '@/src/utils/info';
import { MIGRATION_FLAGS, type MigrationFlags } from '@/src/utils/migration';
import { BASE_FLAGS, type BaseFlags } from '@/src/utils/misc';
import { getSession } from '@/src/utils/session';
import { spinner } from '@/src/utils/spinner';

/**
 * Runs the RONIN command-line interface (CLI) with the provided configuration options.
 * The `@ronin/cli` package intentionally doesn't do this itself, since the CLI is instead
 * automatically installed and exposed via the shorter `ronin` package name.
 *
 * @param config - Options for customizing the behavior of the CLI.
 *
 * @returns Nothing.
 */
export const run = async (config: { version: string }): Promise<void> => {
  let flags: BaseFlags;
  let positionals: Array<string>;

  try {
    ({ values: flags, positionals } = parseArgs({
      args: process.argv,
      options: { ...BASE_FLAGS, ...MIGRATION_FLAGS, ...TYPES_FLAGS },
      strict: true,
      allowPositionals: true,
    }));
  } catch (err) {
    if (err instanceof Error) {
      spinner.fail(err.message);
    }

    process.exit(1);
  }

  // Exit gracefully on SIGINT
  process.on('SIGINT', () => {
    process.exit(0);
  });

  // Exit gracefully on SIGTERM
  process.on('SIGTERM', () => {
    process.exit(0);
  });

  // Flags for printing useful information about the CLI.
  if (flags.help) return printHelp();
  if (flags.version) return printVersion(config.version);

  // This ensures that people can accidentally type uppercase letters and still get the
  // command they are looking for.
  const normalizedPositionals = positionals.map((positional) => positional.toLowerCase());

  // If this environment variable is provided, the CLI will authenticate as an app for a
  // particular space instead of authenticating as an account. This is especially useful
  // in CI, which must be independent of individual people.
  const appToken = process.env.RONIN_TOKEN;

  // If there is no active session, automatically start one and then continue with the
  // execution of the requested sub command, if there is one. If the `login` sub command
  // is invoked, we don't need to auto-login, since the command itself will handle it.
  const session = await getSession();

  if (!(process.stdout.isTTY || session || appToken)) {
    let message = 'If RONIN CLI is invoked from a non-interactive shell, ';
    message +=
      'a `RONIN_TOKEN` environment variable containing an app token must be provided.';

    spinner.fail(message);
    process.exit(1);
  }

  if (!(session || normalizedPositionals.includes('login'))) await logIn(appToken, false);

  // `login` sub command
  if (normalizedPositionals.includes('login')) {
    await logIn(appToken);
    return;
  }

  // `init` sub command
  if (normalizedPositionals.includes('init'))
    return initializeProject(positionals, {
      appToken: appToken,
      sessionToken: session?.token,
    });

  // `diff` sub command
  if (normalizedPositionals.includes('diff')) {
    return diff(appToken, session?.token, flags, positionals);
  }

  // `apply` sub command
  if (normalizedPositionals.includes('apply')) {
    const migrationFilePath = positionals?.[positionals.indexOf('apply') + 1]
      ? path.join(process.cwd(), positionals[positionals.indexOf('apply') + 1])
      : undefined;

    return apply(appToken, session?.token, flags, migrationFilePath);
  }

  // `types` sub command.
  if (normalizedPositionals.includes('types'))
    return generateTypes(appToken, session?.token, flags);

  // `pull` sub command
  if (normalizedPositionals.includes('pull')) {
    return pull(appToken, session?.token, (flags as MigrationFlags).local);
  }

  // If no matching flags or commands were found, render the help, since we don't want to
  // use the main `ronin` command for anything yet.
  return printHelp();
};

export default run;
