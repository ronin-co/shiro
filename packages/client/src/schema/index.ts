export * from 'shiro-syntax/schema';
