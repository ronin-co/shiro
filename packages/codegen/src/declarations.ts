import { SyntaxKind, factory } from 'typescript';

import { genericIdentifiers, identifiers } from '@/src/constants/identifiers';
import { createImportDeclaration } from '@/src/generators/import';

/**
 * ```ts
 * import type { AddQuery, CountQuery, GetQuery, ListQuery, Model, RemoveQuery, SetQuery } from "@ronin/compiler";
 * ```
 */
export const importRoninQueryTypesType = createImportDeclaration({
  identifiers: [
    { name: identifiers.compiler.dmlQueryType.add },
    { name: identifiers.compiler.dmlQueryType.count },
    { name: identifiers.compiler.dmlQueryType.get },
    { name: identifiers.compiler.ddlQueryType.list },
    { name: identifiers.compiler.model },
    { name: identifiers.compiler.dmlQueryType.remove },
    { name: identifiers.compiler.dmlQueryType.set },
  ],
  module: identifiers.compiler.module.root,
  type: true,
});

/**
 * ```ts
 * import type { StoredObject } from "@ronin/compiler";
 * ```
 */
export const importRoninStoredObjectType = createImportDeclaration({
  identifiers: [{ name: identifiers.compiler.storedObject }],
  module: identifiers.compiler.module.root,
  type: true,
});

/**
 * ```ts
 * import type { DeepCallable, ResultRecord } from "@ronin/syntax/queries";
 * ```
 */
export const importSyntaxUtiltypesType = createImportDeclaration({
  identifiers: [
    { name: identifiers.syntax.deepCallable },
    { name: identifiers.syntax.resultRecord },
  ],
  module: identifiers.syntax.module.queries,
  type: true,
});

/**
 * ```ts
 * import type { PromiseTuple, QueryHandlerOptions } from "ronin/types";
 * ```
 */
export const importQueryHandlerOptionsType = createImportDeclaration({
  identifiers: [
    { name: identifiers.ronin.promiseTuple },
    { name: identifiers.ronin.queryHandlerOptions },
  ],
  module: identifiers.ronin.module.types,
  type: true,
});

/**
 * ```ts
 * type ResolveSchema<
 *  TSchema,
 *  TUsing extends Array<string> | 'all',
 *  TKey extends string
 * > = TUsing extends 'all'
 *  ? TSchema
 *  : TKey extends TUsing[number]
 *    ? TSchema
 *    : TSchema extends Array<any>
 *      ? Array<string>
 *      : string;
 * ```
 */
export const resolveSchemaType = factory.createTypeAliasDeclaration(
  undefined,
  identifiers.utils.resolveSchema,
  [
    /**
     * ```ts
     * TSchema
     * ```
     */
    factory.createTypeParameterDeclaration(undefined, genericIdentifiers.schema),

    /**
     * ```ts
     * TUsing extends Array<string> | 'all'
     * ```
     */
    factory.createTypeParameterDeclaration(
      undefined,
      genericIdentifiers.using,
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createUnionTypeNode([
            factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
          ]),
        ]),
        factory.createLiteralTypeNode(
          factory.createStringLiteral(identifiers.utils.all.text),
        ),
      ]),
    ),

    /**
     * ```ts
     * TKey extends string
     * ```
     */
    factory.createTypeParameterDeclaration(
      undefined,
      genericIdentifiers.key,
      factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
    ),
  ],
  factory.createConditionalTypeNode(
    factory.createTypeReferenceNode(genericIdentifiers.using),
    factory.createLiteralTypeNode(
      factory.createStringLiteral(identifiers.utils.all.text),
    ),
    factory.createTypeReferenceNode(genericIdentifiers.schema),

    factory.createConditionalTypeNode(
      factory.createTypeReferenceNode(genericIdentifiers.key),
      factory.createIndexedAccessTypeNode(
        factory.createTypeReferenceNode(genericIdentifiers.using),
        factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
      ),
      factory.createTypeReferenceNode(genericIdentifiers.schema),

      factory.createConditionalTypeNode(
        factory.createTypeReferenceNode(genericIdentifiers.schema),
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
        ]),
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        ]),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
    ),
  ),
);

/**
 * ```ts
 * type JsonPrimitive = string | number | boolean | null;
 * ```
 */
export const jsonPrimitiveType = factory.createTypeAliasDeclaration(
  undefined,
  identifiers.utils.jsonPrimitive,
  undefined,
  factory.createUnionTypeNode([
    factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
    factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
    factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
    factory.createLiteralTypeNode(factory.createNull()),
  ]),
);

/**
 * ```ts
 * type JsonObject = { [key: string]: JsonPrimitive | JsonObject | JsonArray };
 * ```
 */
export const jsonObjectType = factory.createTypeAliasDeclaration(
  undefined,
  identifiers.utils.jsonObject,
  undefined,
  factory.createTypeLiteralNode([
    factory.createIndexSignature(
      undefined,
      [
        factory.createParameterDeclaration(
          undefined,
          undefined,
          factory.createIdentifier('key'),
          undefined,
          factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        ),
      ],
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.utils.jsonPrimitive),
        factory.createTypeReferenceNode(identifiers.utils.jsonObject),
        factory.createTypeReferenceNode(identifiers.utils.jsonArray),
      ]),
    ),
  ]),
);

/**
 * ```ts
 * type JsonArray = Array<JsonPrimitive | JsonObject | JsonArray>;
 * ```
 */
export const jsonArrayType = factory.createTypeAliasDeclaration(
  undefined,
  identifiers.utils.jsonArray,
  undefined,
  factory.createTypeReferenceNode(identifiers.primitive.array, [
    factory.createUnionTypeNode([
      factory.createTypeReferenceNode(identifiers.utils.jsonPrimitive),
      factory.createTypeReferenceNode(identifiers.utils.jsonObject),
      factory.createTypeReferenceNode(identifiers.utils.jsonArray),
    ]),
  ]),
);
