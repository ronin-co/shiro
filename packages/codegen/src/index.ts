import {
  importQueryHandlerOptionsType,
  importRoninQueryTypesType,
  importRoninStoredObjectType,
  importSyntaxUtiltypesType,
  jsonArrayType,
  jsonObjectType,
  jsonPrimitiveType,
  resolveSchemaType,
} from '@/src/declarations';
import { generateModule } from '@/src/generators/module';
import { generateTypes } from '@/src/generators/types';
import { printNodes } from '@/src/utils/print';

import type { Node } from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * Generates the complete `index.d.ts` file for a list of RONIN models.
 *
 * @param models - A list of models to generate the the types for.
 *
 * @returns A string of the complete `index.d.ts` file.
 */
export const generate = (models: Array<Model>): string => {
  // If there is any models that have a `blob()` field, we need to import the
  // `StoredObject` type from the `@ronin/compiler` package.
  const hasStoredObjectFields = models.some((model) =>
    Object.values(model.fields).some((field) => field.type === 'blob'),
  );

  // Each node represents any kind of "block" like
  // an import statement, interface, namespace, etc.
  const nodes = new Array<Node>(
    importRoninQueryTypesType,
    ...(hasStoredObjectFields ? [importRoninStoredObjectType] : []),
    importSyntaxUtiltypesType,
    importQueryHandlerOptionsType,
  );

  // If there is any models that have a `link()` field, we need to add the
  // `ResolveSchemaType` type.
  const hasLinkFields = models.some((model) =>
    Object.values(model.fields).some((field) => field.type === 'link'),
  );
  if (hasLinkFields) nodes.push(resolveSchemaType);

  const hasJsonFields = models.some((model) =>
    Object.values(model.fields).some((field) => field.type === 'json'),
  );
  if (hasJsonFields) nodes.push(jsonArrayType, jsonObjectType, jsonPrimitiveType);

  // Generate and add the type declarations for each model.
  const typeDeclarations = generateTypes(models);

  // Generate and add the `ronin` module augmentation..
  const moduleAugmentation = generateModule(models, typeDeclarations);
  nodes.push(moduleAugmentation);

  return printNodes(nodes);
};
