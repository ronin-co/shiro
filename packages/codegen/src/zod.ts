import { convertToPascalCase } from '@/src/utils/slug';

import type { ModelField } from 'shiro-compiler';

import type { Model } from '@/src/types/model';

type ModelFieldType = Required<ModelField>['type'];

const ZOD_FIELD_TYPES = {
  blob: 'unknown',
  boolean: 'boolean',
  date: 'date', // TODO(@nurodev): Should this be `z.date()` or `z.string().datetime()`?
  json: 'JsonSchema',
  link: 'unknown',
  number: 'number',
  string: 'string',
} satisfies Record<ModelFieldType, string>;

const JSON_SCHEMA = `const JsonLiteralSchema = z.union([
  z.boolean(),
  z.null(),
  z.number(),
  z.string(),
]);

type Json = z.infer<typeof JsonLiteralSchema> | { [key: string]: Json } | Array<Json>;

const ${ZOD_FIELD_TYPES.json}: z.ZodType<Json> = z.lazy(() =>
  z.union([
    JsonLiteralSchema,
    z.array(${ZOD_FIELD_TYPES.json}),
    z.record(${ZOD_FIELD_TYPES.json})
  ])
);
`;

/**
 * Generates the complete `index.ts` Zod schema file for a list of RONIN models.
 *
 * @param models - A list of models to generate the the types for.
 *
 * @returns A string of the complete `index.ts` file.
 */
export const generateZodSchema = (models: Array<Model>): string => {
  const lines = new Array<string | null>('import { z } from "zod";\n');

  // If no models are provided, an empty export is needed to avoid errors.
  if (models.length <= 0) lines.push('export {};');

  // Only add the `JsonSchema` schema if at least one model has a `json` field.
  const hasJsonField = models.some((model) =>
    Object.values(model.fields).some((field) => field.type === 'json'),
  );
  if (hasJsonField) {
    lines.push(JSON_SCHEMA);
  }

  for (const model of models) {
    const modelName = convertToPascalCase(model.slug);

    const entries = new Map<string, string | Map<string, string>>();
    for (const [fieldSlug, field] of Object.entries(model.fields)) {
      const chainedSchemaMethods = new Array<string>();
      switch (field.type) {
        case 'json': {
          chainedSchemaMethods.push(ZOD_FIELD_TYPES.json);
          break;
        }
        default: {
          const fieldType = field.type as ModelFieldType;
          const zodType = ZOD_FIELD_TYPES[fieldType];
          if (!zodType) continue;
          chainedSchemaMethods.push(`z.${zodType}()`);
          break;
        }
      }

      if (fieldSlug === 'email' && field.type === 'string')
        chainedSchemaMethods.push('email()');

      if (field.required !== true && !('defaultValue' in field))
        chainedSchemaMethods.push('optional()');

      if (fieldSlug === 'id' && field.type === 'string')
        chainedSchemaMethods.push('readonly()');

      if (!fieldSlug.includes('.')) {
        entries.set(fieldSlug, chainedSchemaMethods.join('.'));
        continue;
      }

      const [nestedFieldKey, nestedFieldSubKey] = fieldSlug.split('.');
      const nestedFieldMap = entries.get(nestedFieldKey) ?? new Map<string, string>();
      if (typeof nestedFieldMap !== 'string') {
        nestedFieldMap.set(nestedFieldSubKey, chainedSchemaMethods.join('.'));
      }

      entries.set(nestedFieldKey, nestedFieldMap);
    }

    lines.push(
      `export const ${modelName}Schema = z.object({\n${Array.from(entries)
        .map(([key, value]) => {
          if (typeof value === 'string') return `\t${key}: ${value},`;

          const zodObjectBody = Array.from(value.entries())
            .map(([nestedKey, nestedValue]) => `\t\t${nestedKey}: ${nestedValue},`)
            .join('\n');
          return `\t${key}: z.object({\n${zodObjectBody}\n\t}),`;
        })
        .join('\n')}\n});`,
    );
  }

  return `${lines.join('\n')}\n`;
};
