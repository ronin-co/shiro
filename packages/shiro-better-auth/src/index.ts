import {
  add,
  alter,
  batch,
  count,
  create,
  drop,
  get,
  list,
  remove,
  set,
  sql,
  sqlBatch,
} from 'shiro-orm';

import {
  convertWhereClause,
  getModel,
  transformInput,
  transformOrderedBy,
  transformOutput,
} from '@/transform';

import type { Adapter, AdapterInstance } from 'better-auth/types';
import type { ResultRecordBase } from 'shiro-compiler';
import type { createSyntaxFactory } from 'shiro-orm';

type SyntaxFactory = ReturnType<typeof createSyntaxFactory>;

/**
 * Create a new Better Auth adapter to communicate with RONIN.
 *
 * @param [client] - A RONIN syntax factory instance.
 *
 * @returns A Better Auth adapter to communicate with RONIN
 *
 * @example
 * ```ts
 * import { ronin } from 'shiro-better-auth';
 * import { betterAuth } from 'better-auth';
 *
 * const auth = betterAuth({
 *    database: ronin()
 * });
 * ```
 *
 * @example
 * ```ts
 * import { ronin } from 'shiro-better-auth';
 * import { createSyntaxFactory } from 'shiro-orm';
 * import { betterAuth } from 'better-auth';
 *
 * const client = createSyntaxFactory({
 *    token: '1234'
 * });
 * const auth = betterAuth({
 *    database: ronin(client)
 * });
 * ```
 */
export const ronin = (client?: SyntaxFactory): AdapterInstance => {
  return (): Adapter => {
    const factory = client ?? {
      add,
      alter,
      batch,
      count,
      create,
      drop,
      get,
      list,
      remove,
      set,
      sql,
      sqlBatch,
    };

    return {
      id: 'ronin',
      count: async ({ model: slug, where = [] }) => {
        const model = await getModel(factory, slug);
        const instructions = convertWhereClause(where);
        const transformed = await transformInput(instructions, slug);
        return factory.count[model.pluralSlug].with<number>(transformed);
      },
      create: async ({ data, model }) => {
        const transformed = await transformInput(data, model);
        const record = await factory.add[model].with<ResultRecordBase<Date>>(transformed);
        return transformOutput(record);
      },
      delete: async ({ model, where = [] }) => {
        const instructions = convertWhereClause(where);
        await factory.remove[model].with(instructions);
      },
      deleteMany: async ({ model: slug, where = [] }) => {
        const model = await getModel(factory, slug);
        const instructions = convertWhereClause(where);
        const transformed = await transformInput(instructions, slug);
        const results =
          await factory.remove[model.pluralSlug].with<Array<object>>(transformed);
        return results.length;
      },
      findMany: async ({ model: slug, limit, offset, sortBy, where = [] }) => {
        const model = await getModel(factory, slug);
        const instructions = convertWhereClause(where);
        const transformed = await transformInput(instructions, slug);
        const results = (await factory.get[model.pluralSlug]<
          Array<ResultRecordBase<Date>>
        >({
          after: offset?.toString(),
          limitedTo: limit,
          orderedBy: transformOrderedBy(sortBy),
          with: transformed,
        })) as Array<ResultRecordBase<Date>>;

        return results.map((result) => transformOutput(result));
      },
      findOne: async ({ model, where }) => {
        const instructions = convertWhereClause(where);
        const result = await factory.get[model].with<ResultRecordBase<Date> | null>(
          instructions,
        );
        return transformOutput(result);
      },
      update: async ({ model, update, where = [] }) => {
        const instructions = convertWhereClause(where);
        const transformed = await transformInput(update, model);
        const result = await factory.set[model]<ResultRecordBase<Date> | null>({
          with: instructions,
          to: transformed as Record<string, unknown>,
        });
        return transformOutput(result);
      },
      updateMany: async ({ model: slug, update, where = [] }) => {
        const model = await getModel(factory, slug);
        const instructions = convertWhereClause(where);
        const transformed = (await transformInput(update, slug)) as Record<
          string,
          unknown
        >;
        const results = await factory.set[model.pluralSlug]<Array<object>>({
          to: transformed,
          with: instructions,
        });
        return results.length;
      },
    } as Adapter;
  };
};

export default ronin;
