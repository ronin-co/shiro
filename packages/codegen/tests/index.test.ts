import { describe, expect, test } from 'bun:test';
import { blob, model, string } from 'shiro-syntax/schema';

import { generate } from '@/src/index';

describe('generate', () => {
  test('a basic model', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const output = generate([AccountModel]);
    expect(output).toMatchSnapshot();
  });

  test('a basic model with blob field', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
        image: blob(),
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const output = generate([AccountModel]);
    expect(output).toMatchSnapshot();
  });

  test('with no models', () => {
    const output = generate([]);
    expect(output).toMatchSnapshot();
  });
});
