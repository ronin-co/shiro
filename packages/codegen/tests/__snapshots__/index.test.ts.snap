// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generate a basic model 1`] = `
"import type { AddQuery, CountQuery, GetQuery, ListQuery, Model, RemoveQuery, SetQuery } from "shiro-compiler";
import type { DeepCallable, ResultRecord } from "shiro-syntax/queries";
import type { PromiseTuple, QueryHandlerOptions } from "ronin/types";
declare module "ronin" {
    export type Account = ResultRecord & {
        email: string;
        name: string;
    };
    export type Accounts = Array<Account> & {
        moreBefore?: string;
        moreAfter?: string;
    };
    declare const get: {
        /* Get a single account record */
        account: DeepCallable<GetQuery[keyof GetQuery], Account | null>;
        /* Get multiple account records */
        accounts: DeepCallable<GetQuery[keyof GetQuery], Accounts>;
    };
    declare const count: {
        /* Count multiple account records */
        accounts: DeepCallable<CountQuery[keyof CountQuery], number>;
    };
    declare const set: {
        /* Set a single account record */
        account: DeepCallable<SetQuery[keyof SetQuery], Account | null>;
        /* Set multiple account records */
        accounts: DeepCallable<SetQuery[keyof SetQuery], Accounts>;
    };
    declare const add: {
        /* Add a single account record */
        account: DeepCallable<AddQuery[keyof AddQuery], Account | null>;
    };
    declare const remove: {
        /* Remove a single account record */
        account: DeepCallable<RemoveQuery[keyof RemoveQuery], Account | null>;
        /* Remove multiple account records */
        accounts: DeepCallable<RemoveQuery[keyof RemoveQuery], Accounts>;
    };
    declare const batch: <TQueries extends [
        Promise,
        ...Array<Promise>
    ] | Array<Promise>>(operations: () => TQueries, queryOptions?: Record<string, unknown>) => Promise<PromiseTuple<TQueries>>;
    declare const list: {
        /* List all model definitions */
        models: DeepCallable<ListQuery[keyof ListQuery], Array<Model>>;
    };
    declare const createSyntaxFactory: (options: QueryHandlerOptions | (() => QueryHandlerOptions)) => {
        get: typeof get;
        count: typeof count;
        set: typeof set;
        add: typeof add;
        remove: typeof remove;
        list: typeof list;
        create: typeof import("ronin").create;
        alter: typeof import("ronin").alter;
        drop: typeof import("ronin").drop;
        batch: typeof import("ronin").batch;
        sql: typeof import("ronin").sql;
        sqlBatch: typeof import("ronin").sqlBatch;
    };
    export default function (options: QueryHandlerOptions | (() => QueryHandlerOptions)): {
        get: typeof get;
        count: typeof count;
        set: typeof set;
        add: typeof add;
        remove: typeof remove;
        list: typeof list;
        create: typeof import("ronin").create;
        alter: typeof import("ronin").alter;
        drop: typeof import("ronin").drop;
        batch: typeof import("ronin").batch;
        sql: typeof import("ronin").sql;
        sqlBatch: typeof import("ronin").sqlBatch;
    };
}
"
`;

exports[`generate a basic model with blob field 1`] = `
"import type { AddQuery, CountQuery, GetQuery, ListQuery, Model, RemoveQuery, SetQuery } from "shiro-compiler";
import type { StoredObject } from "shiro-compiler";
import type { DeepCallable, ResultRecord } from "shiro-syntax/queries";
import type { PromiseTuple, QueryHandlerOptions } from "ronin/types";
declare module "ronin" {
    export type Account = ResultRecord & {
        email: string;
        image: StoredObject;
        name: string;
    };
    export type Accounts = Array<Account> & {
        moreBefore?: string;
        moreAfter?: string;
    };
    declare const get: {
        /* Get a single account record */
        account: DeepCallable<GetQuery[keyof GetQuery], Account | null>;
        /* Get multiple account records */
        accounts: DeepCallable<GetQuery[keyof GetQuery], Accounts>;
    };
    declare const count: {
        /* Count multiple account records */
        accounts: DeepCallable<CountQuery[keyof CountQuery], number>;
    };
    declare const set: {
        /* Set a single account record */
        account: DeepCallable<SetQuery[keyof SetQuery], Account | null>;
        /* Set multiple account records */
        accounts: DeepCallable<SetQuery[keyof SetQuery], Accounts>;
    };
    declare const add: {
        /* Add a single account record */
        account: DeepCallable<AddQuery[keyof AddQuery], Account | null>;
    };
    declare const remove: {
        /* Remove a single account record */
        account: DeepCallable<RemoveQuery[keyof RemoveQuery], Account | null>;
        /* Remove multiple account records */
        accounts: DeepCallable<RemoveQuery[keyof RemoveQuery], Accounts>;
    };
    declare const batch: <TQueries extends [
        Promise,
        ...Array<Promise>
    ] | Array<Promise>>(operations: () => TQueries, queryOptions?: Record<string, unknown>) => Promise<PromiseTuple<TQueries>>;
    declare const list: {
        /* List all model definitions */
        models: DeepCallable<ListQuery[keyof ListQuery], Array<Model>>;
    };
    declare const createSyntaxFactory: (options: QueryHandlerOptions | (() => QueryHandlerOptions)) => {
        get: typeof get;
        count: typeof count;
        set: typeof set;
        add: typeof add;
        remove: typeof remove;
        list: typeof list;
        create: typeof import("ronin").create;
        alter: typeof import("ronin").alter;
        drop: typeof import("ronin").drop;
        batch: typeof import("ronin").batch;
        sql: typeof import("ronin").sql;
        sqlBatch: typeof import("ronin").sqlBatch;
    };
    export default function (options: QueryHandlerOptions | (() => QueryHandlerOptions)): {
        get: typeof get;
        count: typeof count;
        set: typeof set;
        add: typeof add;
        remove: typeof remove;
        list: typeof list;
        create: typeof import("ronin").create;
        alter: typeof import("ronin").alter;
        drop: typeof import("ronin").drop;
        batch: typeof import("ronin").batch;
        sql: typeof import("ronin").sql;
        sqlBatch: typeof import("ronin").sqlBatch;
    };
}
"
`;

exports[`generate with no models 1`] = `
"import type { AddQuery, CountQuery, GetQuery, ListQuery, Model, RemoveQuery, SetQuery } from "shiro-compiler";
import type { DeepCallable, ResultRecord } from "shiro-syntax/queries";
import type { PromiseTuple, QueryHandlerOptions } from "ronin/types";
declare module "ronin" {
    declare const get: {};
    declare const count: {};
    declare const set: {};
    declare const add: {};
    declare const remove: {};
    declare const batch: <TQueries extends [
        Promise,
        ...Array<Promise>
    ] | Array<Promise>>(operations: () => TQueries, queryOptions?: Record<string, unknown>) => Promise<PromiseTuple<TQueries>>;
    declare const list: {
        /* List all model definitions */
        models: DeepCallable<ListQuery[keyof ListQuery], Array<Model>>;
    };
    declare const createSyntaxFactory: (options: QueryHandlerOptions | (() => QueryHandlerOptions)) => {
        get: typeof get;
        count: typeof count;
        set: typeof set;
        add: typeof add;
        remove: typeof remove;
        list: typeof list;
        create: typeof import("ronin").create;
        alter: typeof import("ronin").alter;
        drop: typeof import("ronin").drop;
        batch: typeof import("ronin").batch;
        sql: typeof import("ronin").sql;
        sqlBatch: typeof import("ronin").sqlBatch;
    };
    export default function (options: QueryHandlerOptions | (() => QueryHandlerOptions)): {
        get: typeof get;
        count: typeof count;
        set: typeof set;
        add: typeof add;
        remove: typeof remove;
        list: typeof list;
        create: typeof import("ronin").create;
        alter: typeof import("ronin").alter;
        drop: typeof import("ronin").drop;
        batch: typeof import("ronin").batch;
        sql: typeof import("ronin").sql;
        sqlBatch: typeof import("ronin").sqlBatch;
    };
}
"
`;
